web-app/src/app/api/stripe/webhook/route.ts

```diff
@@ -0,0 +1,113 @@
+  getFiatTransactionByServicePaymentId,
+  updateFiatTransactionStatus,
+} from "@/lib/db/services/fiatTransaction.service";
+
+const stripe = new Stripe(getEnvSecrets().STRIPE_SECRET_KEY, {
+  apiVersion: "2025-03-31.basil", // Corrected API version
+});
+
+const handleFiatTransactionFailed = async (
+  session: Stripe.Checkout.Session,
+) => {
+  console.log(`🔔  Payment failed for session ${session.id}`);
+
+  const fiatTransaction = await getFiatTransactionByServicePaymentId(
+    session.id,
+  );
+  if (!fiatTransaction) {
+    console.error(`🔔  No fiat transaction found for session ${session.id}`);
+    return;
+  }
+  console.log(`🔔  Updating fiat transaction ${fiatTransaction.id} to FAILED`);
+  await updateFiatTransactionStatus(
+    fiatTransaction.id,
+    FiatTransactionStatus.FAILED,
+  );
+};
+
+const checkSessionPayment = async (session: Stripe.Checkout.Session) => {
+  console.log(`🔔  Check payment for session ${session.id}`);
+  const paymentStatus = session.payment_status;
+  if (paymentStatus !== "paid") {
+    console.error(
+      `🔔  Payment status is ${paymentStatus} for session ${session.id}`,
+    );
+    return;
+  }
+  const fiatTransaction = await getFiatTransactionByServicePaymentId(
+    session.id,
+  );
+  if (!fiatTransaction) {
+    console.error(`🔔  No fiat transaction found for session ${session.id}`);
+    return;
+  }
+  console.log(
+    `🔔  Updating fiat transaction ${fiatTransaction.id} to SUCCEEDED`,
+  );
+  await updateFiatTransactionStatus(
+    fiatTransaction.id,
+    FiatTransactionStatus.SUCCEEDED,
+  );
+};
+
+export async function POST(request: NextRequest) {
+  console.log("🔔  Stripe webhook received");
+
+  const secret = getEnvSecrets().STRIPE_WEBHOOK_SECRET;
+  if (!secret) {
+    return NextResponse.json(
+      { message: "No stripe-webhook-secret found" },
+      { status: 400 },
+    );
+  }
+
+  // Retrieve the event by verifying the signature using the raw body and secret.
+  const signature = request.headers.get("stripe-signature");
+  if (!signature) {
+    return NextResponse.json(
+      { message: "No stripe-signature header found" },
+      { status: 400 },
+    );
+  }
+
+  let event: Stripe.Event | undefined;
+
+  try {
+    event = stripe.webhooks.constructEvent(
+      await request.text(),
+      signature,
+      getEnvSecrets().STRIPE_WEBHOOK_SECRET,
+    );
+  } catch {
+    console.log(`⚠️  Webhook signature verification failed.`);
+    return NextResponse.json(
+      { message: "Webhook signature verification failed" },
+      { status: 400 },
+    );
+  }
+
+  const eventType = event.type;
+  console.log(`⚠️  Webhook received: ${eventType}`);
+
+  const session = event.data.object as Stripe.Checkout.Session;
+  switch (eventType) {
+    case "checkout.session.completed":
+    case "checkout.session.async_payment_succeeded":
+      await checkSessionPayment(session);
+      break;
+    case "checkout.session.expired":
+    case "checkout.session.async_payment_failed":
```

Comment:
I also did not see any manual checks for state updates, or timeouts for failing tx. Are they somewhere?

Reply:
I want to test the hook after deployment to see how it behaves, rather than implementing it based on assumptions. https://github.com/masumi-network/sokosumi/issues/327

Reply:
That's all related to an unsuccessful hook implementation and with this pointing to https://github.com/masumi-network/sokosumi/issues/327
Depending on the outcome new requirements will be made and implemented.

The documentation in the issue links to the general overview of how to work with the webhook. It also includes a [how to handle undellivered events](https://docs.stripe.com/webhooks/process-undelivered-events)

Reply:
If for do bot want to do it here, ok but this is Not a correct implementation from my point of view. It basically has a bug